src/cargo/core/registry.rs

```diff
@@ -260,67 +260,85 @@ impl<'cfg> PackageRegistry<'cfg> {
-                            dep.package_name()
-                        )
-                    })?;
-
-                let source = self
-                    .sources
-                    .get_mut(dep.source_id())
-                    .expect("loaded source not present");
-                let summaries = source.query_vec(dep)?;
-                let (summary, should_unlock) =
-                    summary_for_patch(orig_patch, locked, summaries, source).chain_err(|| {
                         format!(
                             "patch for `{}` in `{}` failed to resolve",
                             orig_patch.package_name(),
                             url,
                         )
                     })?;
-                debug!(
-                    "patch summary is {:?} should_unlock={:?}",
-                    summary, should_unlock
-                );
-                if let Some(unlock_id) = should_unlock {
-                    unlock_patches.push((orig_patch.clone(), unlock_id));
-                }
+                    debug!(
+                        "patch summary is {:?} should_unlock={:?}",
+                        summary, should_unlock
+                    );
+                    if let Some(unlock_id) = should_unlock {
+                        unlock_patches.push((orig_patch.clone(), unlock_id));
+                    }
 
-                if *summary.package_id().source_id().canonical_url() == canonical {
-                    anyhow::bail!(
-                        "patch for `{}` in `{}` points to the same source, but \
+                    if *summary.package_id().source_id().canonical_url() == canonical {
+                        anyhow::bail!(
+                            "patch for `{}` in `{}` points to the same source, but \
                          patches must point to different sources",
-                        dep.package_name(),
-                        url
-                    );
-                }
-                Ok(summary)
-            })
-            .collect::<CargoResult<Vec<_>>>()
-            .chain_err(|| anyhow::format_err!("failed to resolve patches for `{}`", url))?;
+                            dep.package_name(),
+                            url
+                        );
+                    }
+                    Ok(Poll::Ready(summary))
+                })
+                .collect::<CargoResult<Vec<_>>>()
+                .chain_err(|| anyhow::format_err!("failed to resolve patches for `{}`", url))?;
+            if try_summaries.iter().all(|p| p.is_ready()) {
+                break try_summaries
+                    .into_iter()
+                    .map(|p| match p {
+                        Poll::Ready(s) => s,
+                        Poll::Pending => unreachable!("we just check for this!"),
+                    })
+                    .collect::<Vec<_>>();
+            } else {
+                // TODO: dont hot loop for it to be Ready
```

Comment:
FWIW I imagine this bottoms out in something like `self.wait_for_sources_to_be_ready()`. Each source would already be registered in some internal map of `PackageRegistry` if we're blocked on it, and then we'd ask each source, in sequence, "do your blocking thing now".